pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/text/LuceneTextIndexCreator.java

```diff
@@ -0,0 +1,183 @@
+  // TODO: make buffer size configurable choosing a default value based on the heap usage results in design doc
+  private static final int LUCENE_INDEX_MAX_BUFFER_SIZE_MB = 500;
+
+  public static final String LUCENE_INDEX_DOC_ID_COLUMN_NAME = "DocID";
+
+  private final String _textColumn;
+  private final Directory _indexDirectory;
+  private final IndexWriter _indexWriter;
+
+  private int _nextDocId = 0;
+
+  public static final CharArraySet ENGLISH_STOP_WORDS_SET = new CharArraySet(Arrays
+      .asList("a", "an", "and", "are", "as", "at", "be", "but", "by", "for", "if", "in", "into", "is", "it", "no",
+          "not", "of", "on", "or", "such", "that", "the", "their", "then", "than", "there", "these", "they", "this",
+          "to", "was", "will", "with", "those"), true);
+
+  /**
+   * Called by {@link SegmentColumnarIndexCreator}
+   * when building an offline segment. Similar to how it creates per column
+   * dictionary, forward and inverted index, a text index is also created
+   * if text search is enabled on a column.
+   * @param column column name
+   * @param segmentIndexDir segment index directory
+   * @param commit true if the index should be committed (at the end after all documents have
+   *               been added), false if index should not be committed
+   * Note on commit:
+   *               Once {@link SegmentColumnarIndexCreator}
+   *               finishes indexing all documents/rows for the segment, we need to commit and close
+   *               the Lucene index which will internally persist the index on disk, do the necessary
+   *               resource cleanup etc. We commit during {@link DictionaryBasedInvertedIndexCreator#seal()}
+   *               and close during {@link DictionaryBasedInvertedIndexCreator#close()}.
+   *               This lucene index writer is used by both offline and realtime (both during
+   *               indexing in-memory MutableSegment and later during conversion to offline).
+   *               Since realtime segment conversion is again going to go through the offline
+   *               indexing path and will do everything (indexing, commit, close etc), there is
+   *               no need to commit the index from the realtime side. So when the realtime segment
+   *               is destroyed (which is after the realtime segment has been committed and converted
+   *               to offline), we close this lucene index writer to release resources but don't commit.
+   *               This is the reason to have commit flag part of the constructor.
+   */
+  public LuceneTextIndexCreator(String column, File segmentIndexDir, boolean commit) {
+    _textColumn = column;
+    try {
+      // segment generation is always in V1 and later we convert (as part of post creation processing)
+      // to V3 if segmentVersion is set to V3 in SegmentGeneratorConfig.
+      File indexFile = getV1TextIndexFile(segmentIndexDir);
+      _indexDirectory = FSDirectory.open(indexFile.toPath());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource that might leak to a class field, causing a resource leak that slows down or crashes your system. For example, when StandardAnalyzer constructor throws an exception, the *LuceneTextIndexCreator* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *\_indexDirectory* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).